bio/bio.go

```diff
@@ -0,0 +1,323 @@
+// needed. It returns the header of the parser, which is usually parsed prior
+// to the parser being returned by the "NewXXXParser" functions. Unlike the
+// Next() function, Header() can be called as many times as needed. Sometimes
+// files have useful headers, while other times they do not.
+//
+// The following file formats do not have a useful header:
+//
+//	FASTA
+//	FASTQ
+//	Pileup
+//
+// The following file formats do have a useful header:
+//
+//	SLOW5
+func (p *Parser[DataType, DataTypeHeader]) Header() (*DataTypeHeader, error) {
+	return p.parserInterface.Header()
+}
+
+// ParseN returns a countN number of records/reads/lines from the parser.
+func (p *Parser[DataType, DataTypeHeader]) ParseN(countN int) ([]*DataType, error) {
+	var records []*DataType
+	for counter := 0; counter < countN; counter++ {
+		record, err := p.Next()
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				err = nil // EOF not treated as parsing error.
+			}
+			if record != nil {
+				records = append(records, record)
+			}
+			return records, err
+		}
+		records = append(records, record)
+	}
+	return records, nil
+}
+
+// Parse returns all records/reads/lines from the parser, but does not include
+// the header. It can only be called once on a given parser because it will
+// read all the input from the underlying io.Reader before exiting.
+func (p *Parser[DataType, DataTypeHeader]) Parse() ([]*DataType, error) {
+	return p.ParseN(math.MaxInt)
+}
+
+// ParseWithHeader returns all records/reads/lines, plus the header, from the
+// parser. It can only be called once on a given parser because it will read
+// all the input from the underlying io.Reader before exiting.
+func (p *Parser[DataType, DataTypeHeader]) ParseWithHeader() ([]*DataType, *DataTypeHeader, error) {
+	header, headerErr := p.Header()
+	data, err := p.Parse()
+	if headerErr != nil {
+		return data, header, err
+	}
+	if err != nil {
+		return data, header, err
+	}
+	return data, header, nil
+}
+
+/******************************************************************************
+
+Concurrent higher-level functions
+
+******************************************************************************/
+
+// ParseToChannel pipes all records/reads/lines from a parser into a channel,
+// then optionally closes that channel. If parsing a single file,
+// "keepChannelOpen" should be set to false, which will close the channel once
+// parsing is complete. If many files are being parsed to a single channel,
+// keepChannelOpen should be set to true, so that an external function will
+// close channel once all are done parsing.
+//
+// Context can be used to close the parser in the middle of parsing - for
+// example, if an error is found in another parser elsewhere and all files
+// need to close.
+func (p *Parser[DataType, DataTypeHeader]) ParseToChannel(ctx context.Context, channel chan<- DataType, keepChannelOpen bool) error {
```

Comment:
Well, originally, it was to simplify the use of the channel. Ie, you should be able to run `go ParseToChannel` and then when the parsing is done it'll clean up the channel for you. Otherwise, your for loop wouldn't really end when you start receiving from the channel, and you have no way to know if it terminated, unless you put in your own waitgroup or whatever